### PR TITLE
Add non-blocking subscriber

### PR DESCRIPTION
**Pull request by harmrochel**
To be able to use the subscriber functionality of this library while doing some other non-blocking processing I created a new non-blocking subscriber function which takes a callback as additional parameter. This callback is called whenever the original subscriber would block/wait for messages coming in from Redis.

### CODE REVIEW

File: RedisInternal.cpp [hunk 1]
```diff
@@ -164,3 +165,12 @@ std::shared_ptr<RedisObject> RedisObject::parseType(Client &client)
 
     return std::shared_ptr<RedisObject>(new RedisInternalError(RedisInternalError::UnknownType, String(typeChar)));
 }
+
+std::shared_ptr<RedisObject> RedisObject::parseType(Client &client)
+{
+    std::shared_ptr<RedisObject> type = nullptr;
+    while(type==nullptr){
```
Comment (rpj) [inline RedisInternal.cpp:172]:
```suggestion
    while (type==nullptr) {
```

File: RedisInternal.cpp [hunk 2]
```diff
@@ -133,21 +133,22 @@ static TypeParseMap g_TypeParseMap{
...
-        {
-            return std::shared_ptr<RedisObject>(new RedisInternalError(RedisInternalError::Disconnected));
-        }
+        return std::shared_ptr<RedisObject>(new RedisInternalError(RedisInternalError::Disconnected));
+    }
 
-        typeChar = (RedisObject::Type)client.read();
-    } while (typeChar == -1 || typeChar == '\r' || typeChar == '\n');
+    typeChar = (RedisObject::Type)client.read();
+    if(typeChar == -1 || typeChar == '\r' || typeChar == '\n'){
```
Comment (rpj) [inline RedisInternal.cpp:149]:
```suggestion
    if (typeChar == -1 || typeChar == '\r' || typeChar == '\n') {
```

File: RedisInternal.cpp [hunk 3]
```diff
@@ -133,21 +133,22 @@ static TypeParseMap g_TypeParseMap{
     {RedisObject::Type::Error, [](Client &c)
      { return new RedisError(c); }}};
 
-std::shared_ptr<RedisObject> RedisObject::parseType(Client &client)
+std::shared_ptr<RedisObject> RedisObject::parseTypeNonBlocking(Client &client)
 {
-    while (client.connected() && !client.available())
-        ;
+    if(client.connected() && !client.available()) {
```
Comment (rpj) [inline RedisInternal.cpp:138]:
```suggestion
    if (client.connected() && !client.available()) {
```

File: Redis.cpp [hunk 4]
```diff
@@ -233,7 +233,11 @@ RedisSubscribeResult Redis::startSubscribing(RedisMsgCallback messageCallback, R
   subLoopRun = true;
   while (subLoopRun)
   {
-    auto msg = RedisObject::parseType(conn);
+    loopCallback();
+    auto msg = RedisObject::parseTypeNonBlocking(conn);
+    if(msg == nullptr) {
```
Comment (rpj) [inline Redis.cpp:238]:
```suggestion
    if (msg == nullptr) {
```

**Review by rpj — CHANGES_REQUESTED** [review]
Very minor changes requested regarding white space around control structures (to keep the style consistent), but otherwise this design & logic look fantastic! Will approve immediately once these minor changes are addressed (and will publish them as version 2.5.0 accordingly).

My apologies for the delayed response to the PR, I'm still recovering my health and have very little time during the week to focus on this project unfortunately.

I will be sure to approve this PR within 24 hours once you've made these changes, you have my word on that. Thank you a ton @harmrochel, this is very much appreciated!

**Comment by holla2040** [comment]
There needs to be a separate SubscribeNonBlocking example created.

**Comment by harmrochel** [comment]
@rpj I don't need to hit a deadline or something with that PR, so no pressure :D Get well soon!
I applied your changes regarding the formatting and I moved my changes from the Subscribe example to a new SubscribeNonBlocking example as @holla2040 suggested.

**Comment by holla2040** [comment]
I had problems compiling the main branch v2.4.4 of subscribe example, using arduino 1.8.19. I split out msg and err callbacks, which was accepted becoming v2.4.5 today I believe.  I tested the spilt out subscribe version with your non-blocking code and everything worked as expected.

**Review by rpj — APPROVED** [review]

**Comment by rpj** [comment]
@harmrochel The changes look great and I really appreciate the example as well (good catch on that @holla2040 thank you). Excellent work, thank you a ton for contributing and welcome to the team!

PR approved! feel free to merge at your convenience and I will package it up and release it as version 2.5.0 🚀

**Comment by holla2040** [comment]
Does the non blocking subscribe example have msgCallback and errCallback split out? On my phone this evening, will check in the morning.

**Comment by holla2040** [comment]
when using the zip file from https://github.com/harmrochel/arduino-redis.git on a fresh install of arduino 2.0.3. compiling SubscribeNonBlocking.ino gets this error

/home/holla/Arduino/libraries/arduino-redis-subscribe-non-blocking/Redis.h:383:24: note: candidate: 'RedisSubscribeResult Redis::startSubscribingNonBlocking(Redis::RedisMsgCallback, Redis::LoopCallback, Redis::RedisMsgErrorCallback)'
   RedisSubscribeResult startSubscribingNonBlocking(RedisMsgCallback messageCallback, LoopCallback loopCallback, RedisMsgErrorCallback errorCallback = nullptr);
                        ^~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/holla/Arduino/libraries/arduino-redis-subscribe-non-blocking/Redis.h:383:24: note:   no known conversion for argument 1 from 'subscriberLoop(std::function<void()>)::<lambda(Redis*, String, String)>' to 'Redis::RedisMsgCallback' {aka 'void (*)(Redis*, String, String)'}

Here's my 2.4.5 SubscribeNonBlocking that compiles with msgCallback and errCallback as functions.

`
#include <Redis.h>

// this sketch will build for the ESP8266 or ESP32 platform
#ifdef HAL_ESP32_HAL_H_ // ESP32
#include <WiFiClient.h>
#include <WiFi.h>
#else
#ifdef CORE_ESP8266_FEATURES_H // ESP8266
#include <ESP8266WiFi.h>
#endif
#endif

#define WIFI_SSID ""
#define WIFI_PASSWORD ""

#define REDIS_ADDR ""
#define REDIS_PORT 6379
#define REDIS_PASSWORD ""

#define MAX_BACKOFF 300000 // 5 minutes

bool blink;

void setup()
{
  Serial.begin(115200);
  Serial.println();

  WiFi.mode(WIFI_STA);
  WiFi.begin(WIFI_SSID, WIFI_PASSWORD);
  Serial.print("Connecting to the WiFi");
  while (WiFi.status() != WL_CONNECTED)
  {
    delay(250);
    Serial.print(".");
  }
  Serial.println();
  Serial.print("IP Address: ");
  Serial.println(WiFi.localIP());

  auto backoffCounter = -1;
  auto resetBackoffCounter = [&]() {
    backoffCounter = 0;
  };

  pinMode(LED_BUILTIN, OUTPUT);

  resetBackoffCounter();
  while (subscriberLoop(resetBackoffCounter))
  {
    auto curDelay = min((1000 * (int)pow(2, backoffCounter)), MAX_BACKOFF);

    if (curDelay != MAX_BACKOFF)
    {
      ++backoffCounter;
    }

    Serial.printf("Waiting %lds to reconnect...\n", curDelay / 1000);
    delay(curDelay);
  }

  Serial.printf("Done!\n");

}

void msgCallback(Redis *redisInst, String channel, String msg) {
  Serial.printf("Message on channel '%s': \"%s\"\n", channel.c_str(), msg.c_str());
  blink = (msg.c_str()[0] == '1') ? 1 : 0;

  if (channel == "ctrl-close")
  {
    Serial.println("Got message on ctrl-close: ending!");
    redisInst->stopSubscribing();
  }
  else if (channel == "ctrl-add")
  {
    Serial.printf("Adding subscription to channel '%s'\n", msg.c_str());

    if (!redisInst->subscribe(msg.c_str()))
    {
      Serial.println("Failed to add subscription!");
    }
  }
  else if (channel == "ctrl-rm")
  {
    Serial.printf("Removing subscription to channel '%s'\n", msg.c_str());

    if (!redisInst->unsubscribe(msg.c_str()))
    {
      Serial.println("Failed to remove subscription!");
    }
  }
}

void errorCallback(Redis *redisInst, RedisMessageError err) {
  Serial.printf("Subscription error! '%d'\n", err);
}



// returning 'true' indicates the failure was retryable; false is fatal
bool subscriberLoop(std::function<void(void)> resetBackoffCounter)
{
  WiFiClient redisConn;
  if (!redisConn.connect(REDIS_ADDR, REDIS_PORT))
  {
    Serial.println("Failed to connect to the Redis server!");
    return true;
  }

  Redis redis(redisConn);
  auto connRet = redis.authenticate(REDIS_PASSWORD);
  if (connRet == RedisSuccess)
  {
    Serial.println("Connected to the Redis server!");
  }
  else
  {
    Serial.printf("Failed to authenticate to the Redis server! Errno: %d\n", (int)connRet);
    return false;
  }

  redis.subscribe("foo");
  redis.subscribe("bar");

  Serial.println("Listening...");
  resetBackoffCounter();

  auto subRv = redis.startSubscribingNonBlocking(msgCallback, loop, errorCallback);

  redisConn.stop();
  Serial.printf("Connection closed! (%d)\n", subRv);
  return subRv == RedisSubscribeServerDisconnected; // server disconnect is retryable, everything else is fatal
};


void loop() {
  if (blink) {
    digitalWrite(LED_BUILTIN, !digitalRead(LED_BUILTIN));
    delay(50);
  }
}


`

**Comment by holla2040** [comment]
@harmrochel 
redis->publish doesn't seem to work in SubscribeNonBlocking.ino's loop.

**Comment by harmrochel** [comment]
@holla2040 Thanks for the hint, I will have a look at the SubscribeNonBlocking over the weekend.

Regarding the not working publish in the loop: Have you created a new Redis instance to call publish() or have you reused the one which was already used to issue the subscribe command?
The [redis documentation](https://redis.io/commands/subscribe/) says:
> Once the client enters the subscribed state it is not supposed to issue any other commands, except for additional SUBSCRIBE, [SSUBSCRIBE](https://redis.io/commands/ssubscribe), [PSUBSCRIBE](https://redis.io/commands/psubscribe), [UNSUBSCRIBE](https://redis.io/commands/unsubscribe), [SUNSUBSCRIBE](https://redis.io/commands/sunsubscribe), [PUNSUBSCRIBE](https://redis.io/commands/punsubscribe), [PING](https://redis.io/commands/ping), [RESET](https://redis.io/commands/reset) and [QUIT](https://redis.io/commands/quit) commands.

I was able to call publish in the loop with a second Redis instance (which was created with a second WiFiClient instance).

**Comment by holla2040** [comment]
@harmrochel That makes sense, requiring a second connection is fine. I figured it might just work because Redis and RedisInternals have an idle callback pointing main.loop. Other redis clients probably use threads and TCP's signal function call.

**Comment by rpj** [comment]
Released as [version 2.5.0](https://github.com/electric-sheep-co/arduino-redis/releases/tag/2.5.0)! 🚀 

My sincerest thanks again to you @harmrochel (and @holla2040) for this implementation, it's a truly excellent and necessary addition to the library. 🙇

**Comment by holla2040** [comment]
Great to hear, I'm using the branch right now. I'll switch for production this evening. Thanks for all the effort.

**Pull request merged**